src/providers/mssql/qgsmssqlprovider.cpp

```diff
@@ -364,175 +364,186 @@ void QgsMssqlProvider::loadFields()
-          mPrimaryKeyType = PktInt;
-          mPrimaryKeyAttrs << mAttributeFields.size();
-          isIdentity = true;
-        }
-        else if ( sqlTypeName == QLatin1String( "int" ) || sqlTypeName == QLatin1String( "bigint" ) )
-        {
-          pkCandidates << query.value( 3 ).toString();
-        }
+        pkCandidates << query.value( 3 ).toString();
+      }
 
-        if ( sqlType == QVariant::String )
-        {
-          // Field length in chars is column 7 ("Length") of the sp_columns output,
-          // except for uniqueidentifiers which must use column 6 ("Precision").
-          int length = query.value( sqlTypeName.startsWith( QLatin1String( "uniqueidentifier" ), Qt::CaseInsensitive ) ? 6 : 7 ).toInt();
-          if ( sqlTypeName.startsWith( QLatin1Char( 'n' ) ) )
-          {
-            length = length / 2;
-          }
-          mAttributeFields.append(
-            QgsField(
-              colName, sqlType,
-              sqlTypeName,
-              length ) );
-        }
-        else if ( sqlType == QVariant::Double )
-        {
-          mAttributeFields.append(
-            QgsField(
-              colName, sqlType,
-              sqlTypeName,
-              query.value( 6 ).toInt(),
-              sqlTypeName == QLatin1String( "decimal" ) ? query.value( 8 ).toInt() : -1 ) );
-        }
-        else if ( sqlType == QVariant::Date || sqlType == QVariant::DateTime || sqlType == QVariant::Time )
-        {
-          mAttributeFields.append(
-            QgsField(
-              colName, sqlType,
-              sqlTypeName,
-              -1,
-              -1 ) );
-        }
-        else
-        {
-          mAttributeFields.append(
-            QgsField(
-              colName, sqlType,
-              sqlTypeName ) );
-        }
+      // Field nullable
+      const bool nullable = query.value( 10 ).toBool();
 
-        //COLUMN_DEF
-        if ( !query.value( 12 ).isNull() )
+      QgsDebugMsg( QString( "Col name: %1, Nullable: %2" ).arg( colName ).arg( query.value( 10 ).toString() ) );
+
+      QgsFieldConstraints constraints;
+      if ( !nullable )
+        constraints.setConstraint( QgsFieldConstraints::ConstraintNotNull, QgsFieldConstraints::ConstraintOriginProvider );
+
+      if ( sqlType == QVariant::String )
+      {
+        // Field length in chars is column 7 ("Length") of the sp_columns output,
+        // except for uniqueidentifiers which must use column 6 ("Precision").
+        int length = query.value( sqlTypeName.startsWith( QLatin1String( "uniqueidentifier" ), Qt::CaseInsensitive ) ? 6 : 7 ).toInt();
+        if ( sqlTypeName.startsWith( QLatin1Char( 'n' ) ) )
         {
-          mDefaultValues.insert( i, query.value( 12 ).toString() );
+          length = length / 2;
         }
-
-        ++i;
+        QgsField field = QgsField(
+                           colName, sqlType,
+                           sqlTypeName,
+                           length );
+        if ( !nullable )
```

Comment:
Is this check needed? As soon as we add more constraints like unique I guess that's obsolete

Reply:
yes I can remove it